In [1]:
%load_ext sql

In [2]:
%sql mysql+pymysql://root:Jireh56!!@localhost:3306/md_water_services


## View what is in the database

In [3]:
%sql show tables


 * mysql+pymysql://root:***@localhost:3306/md_water_services
9 rows affected.


[('auditor_report',),
 ('data_dictionary',),
 ('employee',),
 ('global_water_access',),
 ('location',),
 ('visits',),
 ('water_quality',),
 ('water_source',),
 ('well_pollution',)]

In [5]:
%%sql
SELECT * 
FROM md_water_services.auditor_report
LIMIT 5;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('SoRu34980', 'well', 1, "Residents admired the official's commitment to enhancing urban life, praising their cooperative and inclusive approach."),
 ('AkRu08112', 'well', 3, "Villagers spoke highly of the official's dedication and genuine interest in their lives, fostering a sense of belonging and appreciation."),
 ('AkLu02044', 'river', 0, "Villagers were touched by the official's interactions, noting their humility, strong work ethic, and respectful attitude."),
 ('AkHa00421', 'well', 3, "Villagers were moved by the official's visit, praising their hard work, humility, and the profound sense of connection they fostered."),
 ('SoRu35221', 'river', 0, "A photographer's lens captures the queue, though his own struggle for water is a hidden part of the story.")]

## So first, grab the location_id and true_water_source_score columns from auditor_report

In [31]:
%%sql
SELECT location_id,true_water_source_score
FROM md_water_services.auditor_report
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('SoRu34980', 1),
 ('AkRu08112', 3),
 ('AkLu02044', 0),
 ('AkHa00421', 3),
 ('SoRu35221', 0)]

## Now, we join the visits table to the auditor_report table. Make sure to grab subjective_quality_score, record_id and location_id.

In [36]:
%%sql
SELECT 
    a.location_id as Auditlocation,
    a.true_water_source_score,
    v.location_id as visits,
    v.record_id
FROM md_water_services.auditor_report as  a
LEFT JOIN md_water_services.visits AS v
    ON a.location_id = v.location_id
LIMIT 10;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('SoRu34980', 1, 'SoRu34980', 5185),
 ('AkRu08112', 3, 'AkRu08112', 59367),
 ('AkLu02044', 0, 'AkLu02044', 37379),
 ('AkHa00421', 3, 'AkHa00421', 51627),
 ('SoRu35221', 0, 'SoRu35221', 28758),
 ('HaAm16170', 1, 'HaAm16170', 31048),
 ('AkRu04812', 3, 'AkRu04812', 1513),
 ('AkRu08304', 3, 'AkRu08304', 1218),
 ('AkRu05107', 2, 'AkRu05107', 8322),
 ('AkRu05215', 3, 'AkRu05215', 21160)]

In [41]:
%%sql SELECT * FROM Visits
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40),
 (3, 'AkLu01628', 'AkLu01628224', datetime.datetime(2021, 1, 1, 9, 53), 1, 0, 1),
 (4, 'AkRu03357', 'AkRu03357224', datetime.datetime(2021, 1, 1, 10, 11), 1, 28, 14)]

Now that we have the record_id for each location, our next step is to retrieve the corresponding scores from the water_quality table. We
are particularly interested in the subjective_quality_score. To do this, we'll JOIN the visits table and the water_quality table, using the
record_id as the connecting key.

In [39]:
%%sql
SELECT 
    a.location_id AS AuditLocation,
    a.true_water_source_score,
    v.location_id AS VisitsLocation,
    v.record_id,
    wq.subjective_quality_score
FROM md_water_services.auditor_report AS a
LEFT JOIN md_water_services.visits AS v
    ON a.location_id = v.location_id
JOIN md_water_services.water_quality AS wq
    ON v.record_id = wq.record_id
LIMIT 10;



 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('SoRu34980', 1, 'SoRu34980', 5185, 1),
 ('AkRu08112', 3, 'AkRu08112', 59367, 3),
 ('AkLu02044', 0, 'AkLu02044', 37379, 0),
 ('AkHa00421', 3, 'AkHa00421', 51627, 3),
 ('SoRu35221', 0, 'SoRu35221', 28758, 0),
 ('HaAm16170', 1, 'HaAm16170', 31048, 1),
 ('AkRu04812', 3, 'AkRu04812', 1513, 3),
 ('AkRu08304', 3, 'AkRu08304', 1218, 3),
 ('AkRu05107', 2, 'AkRu05107', 8322, 2),
 ('AkRu05215', 3, 'AkRu05215', 21160, 10)]

We now drop some duplicate column and rename subjective quality score to surveyor_score and auditor_score to make it clear which score we're looking at in the results set.


In [43]:
%%sql
SELECT 
    a.location_id AS AuditLocation,
    a.true_water_source_score as auditor_score,
    v.location_id AS VisitsLocation,
    v.record_id,
    wq.subjective_quality_score as surveyor_score
FROM md_water_services.auditor_report AS a
LEFT JOIN md_water_services.visits AS v
    ON a.location_id = v.location_id
JOIN md_water_services.water_quality AS wq
    ON v.record_id = wq.record_id
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('SoRu34980', 1, 'SoRu34980', 5185, 1),
 ('AkRu08112', 3, 'AkRu08112', 59367, 3),
 ('AkLu02044', 0, 'AkLu02044', 37379, 0),
 ('AkHa00421', 3, 'AkHa00421', 51627, 3),
 ('SoRu35221', 0, 'SoRu35221', 28758, 0),
 ('HaAm16170', 1, 'HaAm16170', 31048, 1),
 ('AkRu04812', 3, 'AkRu04812', 1513, 3),
 ('AkRu08304', 3, 'AkRu08304', 1218, 3),
 ('AkRu05107', 2, 'AkRu05107', 8322, 2),
 ('AkRu05215', 3, 'AkRu05215', 21160, 10)]

Since were joining 1620 rows of data, we want to keep track of the number of rows we get each time we run a query. We can either set the
maximum number of rows we want from "Limit to 1000 rows" to a larger number like 10000, or we can force SQL to give us all of the results, using
LIMIT 10000.

### The first approcah is for row count

In [46]:
%%sql
SELECT COUNT(*) AS total_rows
FROM md_water_services.auditor_report AS a
LEFT JOIN md_water_services.visits AS v
    ON a.location_id = v.location_id
LEFT JOIN md_water_services.water_quality AS wq
    ON v.record_id = wq.record_id;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(2698,)]

In [86]:
%%sql
SELECT 
    a.location_id AS AuditLocation,
    a.true_water_source_score AS auditor_score,
    v.location_id AS VisitsLocation,
    v.record_id,
    wq.subjective_quality_score AS surveyor_score
FROM md_water_services.auditor_report AS a
LEFT JOIN md_water_services.visits AS v
       ON a.location_id = v.location_id
JOIN md_water_services.water_quality AS wq
       ON v.record_id = wq.record_id
LIMIT 10;



 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('SoRu34980', 1, 'SoRu34980', 5185, 1),
 ('AkRu08112', 3, 'AkRu08112', 59367, 3),
 ('AkLu02044', 0, 'AkLu02044', 37379, 0),
 ('AkHa00421', 3, 'AkHa00421', 51627, 3),
 ('SoRu35221', 0, 'SoRu35221', 28758, 0),
 ('HaAm16170', 1, 'HaAm16170', 31048, 1),
 ('AkRu04812', 3, 'AkRu04812', 1513, 3),
 ('AkRu08304', 3, 'AkRu08304', 1218, 3),
 ('AkRu05107', 2, 'AkRu05107', 8322, 2),
 ('AkRu05215', 3, 'AkRu05215', 21160, 10)]

You got 2698 rows right? Some of the locations were visited multiple times, so these records are duplicated here. To fix it, we set visits.visit_count
= 1 in the WHERE clause. Make sure you reference the alias you used for visits in the join.

In [59]:
%%sql
SELECT 
    a.location_id AS AuditLocation,
    a.true_water_source_score AS auditor_score,
    v.location_id AS VisitsLocation,
    v.record_id,
    wq.subjective_quality_score AS surveyor_score
FROM md_water_services.auditor_report AS a
JOIN md_water_services.visits AS v
       ON a.location_id = v.location_id
JOIN md_water_services.water_quality AS wq
       ON v.record_id = wq.record_id
WHERE v.visit_count = 1
LIMIT 10000;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
1620 rows affected.


[('SoRu34980', 1, 'SoRu34980', 5185, 1),
 ('AkRu08112', 3, 'AkRu08112', 59367, 3),
 ('AkLu02044', 0, 'AkLu02044', 37379, 0),
 ('AkHa00421', 3, 'AkHa00421', 51627, 3),
 ('SoRu35221', 0, 'SoRu35221', 28758, 0),
 ('HaAm16170', 1, 'HaAm16170', 31048, 1),
 ('AkRu04812', 3, 'AkRu04812', 1513, 3),
 ('AkRu08304', 3, 'AkRu08304', 1218, 3),
 ('AkRu05107', 2, 'AkRu05107', 8322, 2),
 ('AkRu05215', 3, 'AkRu05215', 21160, 10),
 ('HaDe16541', 2, 'HaDe16541', 13070, 2),
 ('KiRu30351', 9, 'KiRu30351', 54220, 9),
 ('AmAm09813', 4, 'AmAm09813', 191, 4),
 ('SoRu37005', 0, 'SoRu37005', 51172, 0),
 ('KiRu26067', 3, 'KiRu26067', 21010, 3),
 ('SoRu37100', 0, 'SoRu37100', 40276, 0),
 ('AmDa11611', 1, 'AmDa11611', 38071, 1),
 ('KiRu29290', 3, 'KiRu29290', 7938, 10),
 ('AkRu07801', 3, 'AkRu07801', 25278, 3),
 ('KiZu31015', 3, 'KiZu31015', 47319, 3),
 ('KiIs24026', 2, 'KiIs24026', 2942, 2),
 ('AkRu08480', 2, 'AkRu08480', 47118, 2),
 ('AmBe11043', 3, 'AmBe11043', 53502, 3),
 ('AmRu13718', 9, 'AmRu13718', 31092, 9),
 ('AmRu15209', 3, 'AmRu15209', 42546, 3),
 ('SoRu37428', 3, 'SoRu37428', 25748, 3),
 ('SoRu34787', 2, 'SoRu34787', 30989, 2),
 ('AkRu03204', 3, 'AkRu03204', 36239, 3),
 ('SoMa34137', 3, 'SoMa34137', 4657, 3),
 ('SoRu38874', 3, 'SoRu38874', 17601, 3),
 ('KiMr24320', 2, 'KiMr24320', 9219, 2),
 ('AkRu06460', 2, 'AkRu06460', 42913, 2),
 ('KiHa22748', 9, 'KiHa22748', 43140, 10),
 ('KiRu28673', 1, 'KiRu28673', 57339, 1),
 ('AkHa00193', 9, 'AkHa00193', 25090, 9),
 ('AmAs10732', 0, 'AmAs10732', 53977, 0),
 ('KiRu27585', 3, 'KiRu27585', 2193, 3),
 ('AkRu03541', 2, 'AkRu03541', 23766, 2),
 ('AkRu04448', 3, 'AkRu04448', 55253, 3),
 ('HaRu20155', 0, 'HaRu20155', 56131, 0),
 ('AmAm09990', 5, 'AmAm09990', 3820, 5),
 ('KiZu31024', 0, 'KiZu31024', 53881, 0),
 ('SoIl32392', 3, 'SoIl32392', 2825, 3),
 ('SoRu38699', 3, 'SoRu38699', 13821, 3),
 ('KiAm22079', 9, 'KiAm22079', 50732, 9),
 ('KiRu26316', 3, 'KiRu26316', 27240, 3),
 ('AkKi01456', 9, 'AkKi01456', 57050, 9),
 ('AkRu08152', 2, 'AkRu08152', 43992, 2),
 ('KiRu27444', 7, 'KiRu27444', 18928, 7),
 ('HaRu20693', 1, 'HaRu20693', 55091, 1),
 ('AkRu04981', 3, 'AkRu04981', 8803, 3),
 ('KiRu28580', 1, 'KiRu28580', 14177, 1),
 ('KiRu30008', 1, 'KiRu30008', 25914, 1),
 ('SoMa34631', 2, 'SoMa34631', 55851, 2),
 ('SoKo33227', 1, 'SoKo33227', 31236, 1),
 ('AmRu14465', 3, 'AmRu14465', 23398, 3),
 ('AkHa00008', 3, 'AkHa00008', 57788, 3),
 ('HaRu19826', 9, 'HaRu19826', 55626, 9),
 ('AmDa11947', 1, 'AmDa11947', 23857, 1),
 ('AmPw12680', 3, 'AmPw12680', 45472, 3),
 ('SoRu37841', 6, 'SoRu37841', 18495, 10),
 ('KiRu27884', 1, 'KiRu27884', 33931, 10),
 ('HaDe16514', 1, 'HaDe16514', 42469, 1),
 ('KiMr24591', 0, 'KiMr24591', 44892, 0),
 ('KiIs24014', 1, 'KiIs24014', 36889, 1),
 ('AkRu07523', 3, 'AkRu07523', 22404, 3),
 ('KiZu31170', 9, 'KiZu31170', 17950, 10),
 ('AkRu04753', 9, 'AkRu04753', 55958, 9),
 ('SoRu36774', 2, 'SoRu36774', 35173, 2),
 ('SoRu36239', 2, 'SoRu36239', 43524, 2),
 ('KiRu27776', 3, 'KiRu27776', 23770, 3),
 ('AkRu03164', 3, 'AkRu03164', 42866, 3),
 ('AkRu05809', 2, 'AkRu05809', 8004, 2),
 ('KiRu28632', 5, 'KiRu28632', 32143, 5),
 ('KiRu30389', 4, 'KiRu30389', 17755, 4),
 ('AkKi01375', 3, 'AkKi01375', 40201, 3),
 ('SoRu37530', 0, 'SoRu37530', 55624, 0),
 ('AmBe10959', 4, 'AmBe10959', 23719, 4),
 ('HaRu19828', 2, 'HaRu19828', 25621, 2),
 ('SoIl32999', 2, 'SoIl32999', 32992, 2),
 ('HaRu18571', 2, 'HaRu18571', 56000, 2),
 ('AkRu08666', 6, 'AkRu08666', 56276, 6),
 ('AkRu03306', 7, 'AkRu03306', 29135, 7),
 ('AkRu08936', 3, 'AkRu08936', 132, 3),
 ('HaSe21341', 3, 'HaSe21341', 59324, 3),
 ('SoRu35327', 1, 'SoRu35327', 59821, 1),
 ('AmRu14903', 9, 'AmRu14903', 54574, 9),
 ('AmAb09101', 6, 'AmAb09101', 10635, 6),
 ('KiAm22276', 0, 'KiAm22276', 15269, 0),
 ('HaRu20617', 1, 'HaRu20617', 18740, 1),
 ('KiZu31370', 3, 'KiZu31370', 36864, 10),
 ('HaRu18217', 1, 'HaRu18217', 58386, 1),
 ('KiRu29806', 9, 'KiRu29806', 35488, 9),
 ('KiRu28057', 1, 'KiRu28057', 25822, 1),
 ('AkRu06170', 3, 'AkRu06170', 16253, 3),
 ('AkRu03380', 3, 'A

So, to do this, we need to grab the type_of_water_source column from the water_source table and call it survey_source, using the
source_id column to JOIN. Also select the type_of_water_source from the auditor_report table, and call it auditor_source.

### To detect actual errors or mismatches, we can take it one step further:

In [58]:
%%sql
SELECT 
    a.location_id,
    COUNT(DISTINCT ws.type_of_water_source) AS unique_source_types
FROM md_water_services.auditor_report AS a
LEFT JOIN md_water_services.visits AS v
       ON a.location_id = v.location_id
JOIN md_water_services.water_source AS ws
       ON v.source_id = ws.source_id
WHERE v.visit_count = 1
GROUP BY a.location_id
HAVING unique_source_types > 1
LIMIT 10000;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
0 rows affected.


[]

In [74]:
%%sql
WITH Incorrect_records AS (
    SELECT 
        a.location_id,
        a.type_of_water_source AS auditor_source,
        ws.type_of_water_source AS surveyor_source,
        v.visit_count
    FROM md_water_services.auditor_report AS a
    LEFT JOIN md_water_services.visits AS v
           ON a.location_id = v.location_id
    JOIN md_water_services.water_source AS ws
           ON v.source_id = ws.source_id
    WHERE v.visit_count = 1
)
SELECT *
FROM Incorrect_records
limit 5;



 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('SoRu34980', 'well', 'well', 1),
 ('AkRu08112', 'well', 'well', 1),
 ('AkLu02044', 'river', 'river', 1),
 ('AkHa00421', 'well', 'well', 1),
 ('SoRu35221', 'river', 'river', 1)]

### Get a unique list of employees


In [75]:
%%sql
WITH Incorrect_records AS (
    SELECT
        auditor_report.location_id,
        visits.record_id,
        emp.employee_name,
        auditor_report.true_water_source_score AS auditor_score,
        wq.subjective_quality_score AS surveyor_score,
        auditor_report.statements
    FROM auditor_report
    JOIN visits
        ON auditor_report.location_id = visits.location_id
    JOIN water_quality wq
        ON visits.record_id = wq.record_id
    JOIN employee emp
        ON emp.assigned_employee_id = visits.assigned_employee_id
    WHERE visits.visit_count = 1
      AND auditor_report.true_water_source_score != wq.subjective_quality_score
),
Unique_employees AS (
    SELECT DISTINCT employee_name
    FROM Incorrect_records
)
SELECT *
FROM Unique_employees
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('Rudo Imani',),
 ('Bello Azibo',),
 ('Zuriel Matembo',),
 ('Yewande Ebele',),
 ('Jengo Tumaini',)]

## Count the number of mistakes

In [67]:
%%sql
WITH Incorrect_records AS (
    SELECT
        auditor_report.location_id,
        visits.record_id,
        emp.employee_name,
        auditor_report.true_water_source_score AS auditor_score,
        wq.subjective_quality_score AS surveyor_score,
        auditor_report.statements
    FROM auditor_report
    JOIN visits
        ON auditor_report.location_id = visits.location_id
    JOIN water_quality wq
        ON visits.record_id = wq.record_id
    JOIN employee emp
        ON emp.assigned_employee_id = visits.assigned_employee_id
    WHERE visits.visit_count = 1
      AND auditor_report.true_water_source_score != wq.subjective_quality_score
),
error_count AS (
    SELECT
        employee_name,
        COUNT(*) AS number_of_mistakes
    FROM Incorrect_records
    GROUP BY employee_name
)
SELECT *
FROM error_count
ORDER BY number_of_mistakes DESC
LIMIT 5;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('Bello Azibo', 26),
 ('Malachi Mavuso', 21),
 ('Zuriel Matembo', 17),
 ('Lalitha Kaburi', 7),
 ('Rudo Imani', 5)]

### Save Incorrect_records as a VIEW

In [68]:
%%sql
CREATE OR REPLACE VIEW Incorrect_records AS
SELECT
    auditor_report.location_id,
    visits.record_id,
    emp.employee_name,
    auditor_report.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score,
    auditor_report.statements
FROM auditor_report
JOIN visits
    ON auditor_report.location_id = visits.location_id
JOIN water_quality wq
    ON visits.record_id = wq.record_id
JOIN employee emp
    ON emp.assigned_employee_id = visits.assigned_employee_id
WHERE visits.visit_count = 1
  AND auditor_report.true_water_source_score != wq.subjective_quality_score;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
0 rows affected.


[]

## Count mistakes per employee (error_count)

In [71]:
%%sql
WITH error_count AS (
    SELECT
        employee_name,
        COUNT(*) AS number_of_mistakes
    FROM Incorrect_records
    GROUP BY employee_name
)
SELECT * 
FROM error_count
ORDER BY number_of_mistakes DESC
LIMIT 5;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('Bello Azibo', 26),
 ('Malachi Mavuso', 21),
 ('Zuriel Matembo', 17),
 ('Lalitha Kaburi', 7),
 ('Rudo Imani', 5)]

## Find the average mistakes

In [73]:
%%sql
WITH error_count AS (
    SELECT
        employee_name,
        COUNT(*) AS number_of_mistakes
    FROM Incorrect_records
    GROUP BY employee_name
)
SELECT AVG(number_of_mistakes) AS avg_error_count_per_empl
FROM error_count;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(Decimal('6.0000'),)]

### Identify suspect employees (above average)

In [76]:
%%sql
WITH error_count AS (
    SELECT
        employee_name,
        COUNT(*) AS number_of_mistakes
    FROM Incorrect_records
    GROUP BY employee_name
),
suspect_list AS (
    SELECT
        employee_name,
        number_of_mistakes
    FROM error_count
    WHERE number_of_mistakes > (SELECT AVG(number_of_mistakes) FROM error_count)
)
SELECT * 
FROM suspect_list
ORDER BY number_of_mistakes DESC;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
4 rows affected.


[('Bello Azibo', 26),
 ('Malachi Mavuso', 21),
 ('Zuriel Matembo', 17),
 ('Lalitha Kaburi', 7)]

### Create a view of all mismatched records between auditors and surveyors

In [77]:
%%sql
CREATE OR REPLACE VIEW Incorrect_records AS
SELECT
    auditor_report.location_id,
    visits.record_id,
    emp.employee_name,
    auditor_report.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score,
    auditor_report.statements
FROM auditor_report
JOIN visits
    ON auditor_report.location_id = visits.location_id
JOIN water_quality wq
    ON visits.record_id = wq.record_id
JOIN employee emp
    ON emp.assigned_employee_id = visits.assigned_employee_id
WHERE visits.visit_count = 1
  AND auditor_report.true_water_source_score != wq.subjective_quality_score;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
0 rows affected.


[]

## we now extend the error_count CTE and calculate the average number of mistakes, then use it as a filter to find employees above that threshold.

In [79]:
%%sql
WITH error_count AS (
    SELECT
        employee_name,
        COUNT(*) AS number_of_mistakes
    FROM Incorrect_records
    GROUP BY employee_name
)
# select employees above the average number of mistakes
SELECT
    employee_name,
    number_of_mistakes
FROM error_count
WHERE number_of_mistakes > (SELECT AVG(number_of_mistakes) FROM error_count)
ORDER BY number_of_mistakes DESC;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
4 rows affected.


[('Bello Azibo', 26),
 ('Malachi Mavuso', 21),
 ('Zuriel Matembo', 17),
 ('Lalitha Kaburi', 7)]

## Filter Incorrect_records to only show the suspects’ records, including their statements.

In [84]:
%%sql
WITH error_count AS (
    SELECT
        employee_name,
        COUNT(*) AS number_of_mistakes
    FROM Incorrect_records
    GROUP BY employee_name
),

-- Step 2: suspect_list = employees with above-average mistakes
suspect_list AS (
    SELECT
        employee_name,
        number_of_mistakes
    FROM error_count
    WHERE number_of_mistakes > (SELECT AVG(number_of_mistakes) FROM error_count)
)

-- Step 3: Pull all mismatched records (with statements) for suspects
SELECT
    ir.employee_name,
    ir.location_id,
    ir.auditor_score,
    ir.surveyor_score,
    ir.statements
FROM Incorrect_records ir
WHERE ir.employee_name IN (SELECT employee_name FROM suspect_list)
ORDER BY ir.employee_name, ir.location_id
LIMIT 10;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('Bello Azibo', 'AkHa00363', 0, 10, "A local mason's strong back is bent by the wait for water, his concerns for quality reflecting a broader challenge."),
 ('Bello Azibo', 'AkLu02500', 3, 10, "A tour guide's enthusiasm is dampened by the water situation, his own wait in line a daily challenge."),
 ('Bello Azibo', 'AkRu03848', 3, 10, "A local tailor's creativity is stifled by frustration, his work impacted by the corruption that taints the water."),
 ('Bello Azibo', 'AkRu04508', 1, 10, 'An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.'),
 ('Bello Azibo', 'AkRu05741', 3, 10, 'An air of mistrust surrounded the official, as villagers spoke of laziness and hints of corruption. The mention of cash passing discreetly only deepened their concerns.'),
 ('Bello Azibo', 'AkRu06495', 2, 10, 'A healthcare worker in the queue expressed fears about water-borne diseases, her face etched with worry.","'),
 ('Bello Azibo', 'AmRu14364', 2, 10, "A local musician's songs of joy are replaced with melodies of despair, inspired by the corrupt who control the water."),
 ('Bello Azibo', 'AmRu14691', 2, 10, "An artisan's creative mind is stifled by the wait, thoughts of craft overshadowed by concern for water quality."),
 ('Bello Azibo', 'AmRu14842', 2, 10, "An elderly woman's weary eyes reflect the toll of daily hours spent in the queue, her thoughts on family back home."),
 ('Bello Azibo', 'AmRu14887', 9, 10, 'Villagers expressed their discomfort with an official who displayed a haughty demeanor and negligence. The mention of cash transactions deepened their growing sense of unease.')]

## Filter suspect records mentioning “cash”

In [85]:
%%sql
WITH error_count AS ( -- Count mistakes per employee
    SELECT
        employee_name,
        COUNT(*) AS number_of_mistakes
    FROM Incorrect_records
    GROUP BY employee_name
),
suspect_list AS ( -- Employees with above-average mistakes
    SELECT
        employee_name,
        number_of_mistakes
    FROM error_count
    WHERE number_of_mistakes > (SELECT AVG(number_of_mistakes) FROM error_count)
)
-- Now filter records where statements mention "cash"
SELECT
    ir.employee_name,
    ir.location_id,
    ir.statements
FROM Incorrect_records ir
WHERE ir.employee_name IN (SELECT employee_name FROM suspect_list)
  AND ir.statements LIKE '%cash%'
ORDER BY ir.employee_name, ir.location_id;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
19 rows affected.


[('Bello Azibo', 'AkRu04508', 'An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.'),
 ('Bello Azibo', 'AkRu05741', 'An air of mistrust surrounded the official, as villagers spoke of laziness and hints of corruption. The mention of cash passing discreetly only deepened their concerns.'),
 ('Bello Azibo', 'AmRu14887', 'Villagers expressed their discomfort with an official who displayed a haughty demeanor and negligence. The mention of cash transactions deepened their growing sense of unease.'),
 ('Bello Azibo', 'HaSe21323', 'Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.'),
 ('Bello Azibo', 'KiIs23853', "Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The mention of cash changing hands further tainted their perception."),
 ('Bello Azibo', 'KiMr24919', "Suspicion and unease colored the villagers' accounts of an official's haughty behavior and potential corruption. The mention of cash changing hands added to their apprehension."),
 ('Bello Azibo', 'KiRu27065', 'Villagers expressed their discomfort with an official who displayed a haughty demeanor and negligence. The mention of cash transactions deepened their growing sense of unease.'),
 ('Bello Azibo', 'KiRu29639', "An unsettling atmosphere prevailed as villagers shared stories of an official's arrogance and perceived corruption. The mention of cash exchanges only intensified their concerns."),
 ('Lalitha Kaburi', 'AkRu07310', 'Villagers spoke of their unsettling encounters with an official who seemed indifferent and uninterested, hinting at potential improprieties involving cash exchanges.'),
 ('Lalitha Kaburi', 'KiRu29329', "Suspicion colored villagers' descriptions of an official's aloof demeanor and apparent laziness. The reference to cash transactions cast doubt on their motives."),
 ('Malachi Mavuso', 'AmAm09607', 'Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.'),
 ('Malachi Mavuso', 'AmAm09956', 'An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.'),
 ('Malachi Mavuso', 'AmRu15719', "Suspicion and unease colored the villagers' accounts of an official's haughty behavior and potential corruption. The mention of cash changing hands added to their apprehension."),
 ('Malachi Mavuso', 'KiRu25347', 'Villagers expressed their discontent with an official who appeared dismissive and neglectful. The mention of cash changing hands added to their growing sense of distrust.'),
 ('Zuriel Matembo', 'AkRu05880', "Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The allusion to cash changing hands deepened their skepticism."),
 ('Zuriel Matembo', 'HaRu20146', "An air of mistrust hung heavy as villagers described an official's aloof behavior and potential impropriety. The whisper of cash exchanges only fueled their apprehension."),
 ('Zuriel Matembo', 'HaSe20888', "Suspicion and unease colored the villagers' accounts of an official's haughty behavior and potential corruption. The mention of cash changing hands added to their apprehension."),
 ('Zuriel Matembo', 'SoIl32575', 'Villagers recounted unsettling encounters with an official known for their arrogance and avoidance of responsibilities. The mention of cash changing hands added to their apprehension and distrust.'),
 ('Zuriel Matembo', 'SoRu38331', 'An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only i

### Double-check: are there any non-suspects accused of “cash”?

In [83]:
%%sql
WITH error_count AS (
    SELECT
        employee_name,
        COUNT(*) AS number_of_mistakes
    FROM Incorrect_records
    GROUP BY employee_name
),
suspect_list AS (
    SELECT
        employee_name
    FROM error_count
    WHERE number_of_mistakes > (SELECT AVG(number_of_mistakes) FROM error_count)
)
SELECT
    ir.employee_name,
    ir.location_id,
    ir.statements
FROM Incorrect_records ir
WHERE ir.employee_name NOT IN (SELECT employee_name FROM suspect_list)
  AND ir.statements LIKE '%cash%';


 * mysql+pymysql://root:***@localhost:3306/md_water_services
0 rows affected.


[]